meta-lxatac-bsp/recipes-kernel/linux/files/patches/0306-Revert-drm-format-helper-Cache-buffers-with-struct-d.patch

```diff
@@ -0,0 +1,259 @@
+- * temporary storage.
+- */
+-void drm_format_conv_state_copy(struct drm_format_conv_state *state,
+-				const struct drm_format_conv_state *old_state)
+-{
+-	/*
+-	 * So far, there's only temporary storage here, which we don't
+-	 * duplicate. Just clear the fields.
+-	 */
+-	state->tmp.mem = NULL;
+-	state->tmp.size = 0;
+-	state->tmp.preallocated = false;
+-}
+-EXPORT_SYMBOL(drm_format_conv_state_copy);
+-
+-/**
+- * drm_format_conv_state_reserve - Allocates storage for format conversion
+- * @state: The format-conversion state
+- * @new_size: The minimum allocation size
+- * @flags: Flags for kmalloc()
+- *
+- * Allocates at least @new_size bytes and returns a pointer to the memory
+- * range. After calling this function, previously returned memory blocks
+- * are invalid. It's best to collect all memory requirements of a format
+- * conversion and call this function once to allocate the range.
+- *
+- * Returns:
+- * A pointer to the allocated memory range, or NULL otherwise.
+- */
+-void *drm_format_conv_state_reserve(struct drm_format_conv_state *state,
+-				    size_t new_size, gfp_t flags)
+-{
+-	void *mem;
+-
+-	if (new_size <= state->tmp.size)
+-		goto out;
+-	else if (state->tmp.preallocated)
+-		return NULL;
+-
+-	mem = krealloc(state->tmp.mem, new_size, flags);
+-	if (!mem)
+-		return NULL;
+-
+-	state->tmp.mem = mem;
+-	state->tmp.size = new_size;
+-
+-out:
+-	return state->tmp.mem;
+-}
+-EXPORT_SYMBOL(drm_format_conv_state_reserve);
+-
+-/**
+- * drm_format_conv_state_release - Releases an format-conversion storage
+- * @state: The format-conversion state
+- *
+- * Releases the memory range references by the format-conversion state.
+- * After this call, all pointers to the memory are invalid. Prefer
+- * drm_format_conv_state_init() for cleaning up and unloading a driver.
+- */
+-void drm_format_conv_state_release(struct drm_format_conv_state *state)
+-{
+-	if (state->tmp.preallocated)
+-		return;
+-
+-	kfree(state->tmp.mem);
+-	state->tmp.mem = NULL;
+-	state->tmp.size = 0;
+-}
+-EXPORT_SYMBOL(drm_format_conv_state_release);
+-
+ static unsigned int clip_offset(const struct drm_rect *clip, unsigned int pitch, unsigned int cpp)
+ {
+ 	return clip->y1 * pitch + clip->x1 * cpp;
+@@ -136,7 +45,6 @@ EXPORT_SYMBOL(drm_fb_clip_offset);
+ static int __drm_fb_xfrm(void *dst, unsigned long dst_pitch, unsigned long dst_pixsize,
+ 			 const void *vaddr, const struct drm_framebuffer *fb,
+ 			 const struct drm_rect *clip, bool vaddr_cached_hint,
+-			 struct drm_format_conv_state *state,
+ 			 void (*xfrm_line)(void *dbuf, const void *sbuf, unsigned int npixels))
+ {
+ 	unsigned long linepixels = drm_rect_width(clip);
+@@ -152,7 +60,7 @@ static int __drm_fb_xfrm(void *dst, unsigned long dst_pitch, unsigned long dst_p
+ 	 * one line at a time.
+ 	 */
+ 	if (!vaddr_cached_hint) {
+-		stmp = drm_format_conv_state_reserve(state, sbuf_len, GFP_KERNEL);
++		stmp = kmalloc(sbuf_len, GFP_KERNEL);
```

Comment:
Did you report this upstream?

Reply:
Not yet. If I go through the stress of dealing with the mailing list I'd like to provide some more information than just "this leaks memory now". But I did not get around to looking at it in enough depth to provide this information.

We may also already have a proper solution now. @lynxeye-dev sent me a patch to test, but I did not get around to doing that either, but that's what I will do next.

Reply:
I've tested @lynxeye-dev's patch now and confirmed that it does indeed work. I'am confident that he will send it upstream soon.

I've updated the PR now to contain the patch instead of the large revert.